### DISCUSSION TITLE
2038 support and allowing for testing 2038 support within DOSEMU2

### DISCUSSION THREAD

**Discussion by jschwartzenberg** [dosemu2/dosemu2#2221] [Ideas]
https://linux.slashdot.org/story/24/06/20/2041244/suse-upgrades-its-distros-with-19-years-of-support

I came across this article and came up with the idea that maybe DOSEMU2 could feature a mode in which the timestamps visible to DOS applications would be beyond 2038. That way we could already start testing what the DOS world will look like once we reach 2038.

I suspect the world beyond 2038 will cause a revolution in retrocomputing. Maybe DOSEMU2 could start it :)

**Comment by stsp** [comment]
https://news.ycombinator.com/item?id=25311468
```
On the other hand, MS-DOS, which internally uses a 1980 year-epoch with a byte-sized year counter (and does not
use a seconds-based epoch unlike Unixes), will be fine up to 2107
```
Neither we, nor dosemu2, nor DOS is
to survive up to 2107.

**Reply by jschwartzenberg** [reply]
Yeah I'm afraid you're right about that, but if somehow I do make it to 2107, I'll be sure to remember this :P

Btw. what about applications running on DOS? Do you think they might still have 2038 limitations or would every single DOS application likely work between 2038 and 2107?

**Reply by stsp** [reply]
I _think_ (a totally wild guess) that
most of DOS apps have an Y2K problem,
as they were already quite dead back
then, to be updated. So they likely already
truncate the date.
But other than that, I am pretty sure
they have no 2038 limitation.
@mceric may know more about that,
as I am really not interested in DOS apps
at all (oops! :), I am just a hypervisor writer.

Also please note that for DOS apps all
these problems are not fatal, they are
only fatal for DOS itself. For an app that
qould only mean showing the wrong/truncated
date - a big deal?

> but if somehow I do make it to 2107, I'll be sure to remember this :P

Even if you do, you'll have problems
remembering anything, you'll have
problems recalling what DOS is, and
I am pretty sure there will be no DOS
and/or DOS apps around, neither its
emulators. They all will be outdated,
unbuildable, incompatible with the
existing OSes, irrelevant, and so on.

**Reply by stsp** [reply]
@mceric mentions (privately as usual)
that DOS apps are likely to have problems
starting from 2070, since that would
sign-overflow "days since 01jan1980"
if you use the 2-byte signed word as
a counter. And there may be additional
complication passing by 2100.
Well, at least some of us can try to reach
2070, but will they be able to recall this
discussion, is very questionable. :)

**Reply by jschwartzenberg** [reply]
You do keep putting up challenges here :)

**Comment by jharrison03** [comment]
I think this "problem" borders on philosophy.  If I'm still around using DOS stuff in 2038, I would set the clock black and carry on.

**Comment by stsp** [comment]
In fact, after watching the last
Windows-11 update on youtube,
I think in 2107 (not so much in
2038) you'll be able to play any
DOS game by just thinking about
it. Or, at worst, by saying something.
The emulator will be produced by
AI transparently, and any bugs in
a game (including the date bug)
will just be patched by AI in a second.
I don't think AI will have any problems
decompiling any game, patching it
and interpreting directly from the
decompiled sources. :)

**Reply by stsp** [reply]
Or, more simply, it could make you
think you've just played that game. :)
Then no efforts are needed at all.

**Comment by stsp** [comment]
In fact, speaking of AI, we could
really get some use of the AI-based
renderers.
Just imagine some Test Drive 2 from
80th, with AI drawing all the missing
details in a background, in HD res...
That would be very funny I suppose. :)

**Reply by jschwartzenberg** [reply]
Yeah I was about to say, should we really create another topic on AI integration for DOSEMU2?

**Reply by stsp** [reply]
Well, perhaps you can create such
a topic on libsdl forums. If they provide
an API for AI-based renderers, maybe
we can use those.
Of course if you have some other AI-based
idea, you can try also here. :)

**Reply by stsp** [reply]
Actually it appears this method was
already tried:
https://www.youtube.com/watch?v=2y1j3qlY0rg
As you can see, all the AI can do
for old games right now, is to
"modernize" the screen-shots.
And it completely messes up the
text, if it is anywhere on a screen.
At the end of the video the guy says
that maybe there will soon be a
program to "modernize" videos, and
not just screen-shots.
So it seems, all the AI can now do
properly, is the texture upscaling.
Which is likely not what dosemu2
could use.

**Reply by stuaxo** [reply]
Moving images that stay coherent with AI is a whole other set of problems.
